File: codes/84. 柱状图中最大的矩形.java
/*
方法一：时间复杂度为On2的遍历，容易超时
 */
class Solution {
    public int largestRectangleArea(int[] heights) {
        int max_area = 0;
        for(int i=0; i<heights.length; i++) {
            int len = 1;
            for(int j=i-1; j>=0 && heights[j] >= heights[i]; j--) {
                len++;
            }
            for(int j=i+1; j<heights.length && heights[j] >= heights[i]; j++) {
                len++;
            }
            max_area = Math.max(max_area, len*heights[i]);
        }
        return max_area;
    }
}

/*
  方法二：使用单调栈。单调栈可以找出当前元素左侧最近的比该元素小/大的元素
 */
class Solution {
    static class Rectangle{
        int index;
        int height;
        Rectangle(int index, int height) {
            this.index = index;
            this.height = height;
        }
    }
    public int largestRectangleArea(int[] heights) {
        ArrayList<Rectangle> stack = new ArrayList<>();
        int[] left_index = new int[heights.length];
        int max_area = 0;

        // find boundary
        for(int i=0; i<heights.length; i++) {
            int height = heights[i];
            while (stack.size()>0 && stack.get(stack.size()-1).height>=height){
                Rectangle removed = stack.remove(stack.size()-1);
                max_area = Math.max(max_area, heights[removed.index] * (i-left_index[removed.index]));
            }
            if(stack.size()==0) left_index[i] = 0;
            else left_index[i] = stack.get(stack.size()-1).index+1;
            stack.add(new Rectangle(i, height));
        }
        while (stack.size()>0) {
            Rectangle removed = stack.remove(stack.size()-1);
            max_area = Math.max(max_area, heights[removed.index] * (heights.length-left_index[removed.index]));
        }
        return max_area;
    }
}

File: codes/84. 柱状图中最大的矩形.go
func largestRectangleArea(heights []int) int {
	n := len(heights)
	areas := make([]int, n)
	monoStack := make([]int, 0)
	for i := 0; i < n; i++ {
		leftbound := -1
		for len(monoStack) > 0 && heights[monoStack[len(monoStack)-1]] >= heights[i] {
			monoStack = monoStack[:len(monoStack)-1]
		}
		if len(monoStack) > 0 {
			leftbound = monoStack[len(monoStack)-1]
		}
		monoStack = append(monoStack, i)
		areas[i] = (i - leftbound) * heights[i]
	}
	monoStack = make([]int, 0)
	for i := n - 1; i >= 0; i-- {
		rightbound := n
		for len(monoStack) > 0 && heights[monoStack[len(monoStack)-1]] >= heights[i] {
			monoStack = monoStack[:len(monoStack)-1]
		}
		if len(monoStack) > 0 {
			rightbound = monoStack[len(monoStack)-1]
		}
		monoStack = append(monoStack, i)
		areas[i] += (rightbound - i - 1) * heights[i]
	}
	maxarea := 0
	for _, area := range areas {
		if area > maxarea {
			maxarea = area
		}
	}
	return maxarea
}